examples/simple.rs

```diff
@@ -1,40 +1,65 @@
-
 extern crate rspec;
-use rspec::context::describe;
+
 use std::io;
+use std::sync::{Arc, Mutex};
+use std::collections::BTreeSet;
 
-pub fn main() {
-    let stdout = &mut io::stdout();
-    let mut formatter = rspec::formatter::Simple::new(stdout);
-    let mut runner = describe("rspec is a classic BDD testing", |ctx| {
+use rspec::prelude::*;
 
-        ctx.it("can define tests", || true);
+pub fn main() {
+    let simple = rspec::formatter::Simple::new(io::stdout());
+    let formatter = Arc::new(Mutex::new(simple));
+    let configuration = Configuration::default().parallel(false);
+    let runner = Runner::new(configuration, vec![formatter]);
 
-        ctx.describe("rspec use results for tests results", |ctx| {
+    #[derive(Clone, Debug)]
+    struct Environment {
+        set: BTreeSet<usize>,
+        len_before: usize,
+    }
 
-            ctx.it("passed if the return is_ok()", || Ok(()) as Result<(),()>);
+    let environment = Environment {
+        set: BTreeSet::new(),
+        len_before: 0,
+    };
 
-            ctx.it("failed if the return is_err()", || Err(()) as Result<(),()>);
+    runner.run_or_exit(rspec::given("a BTreeSet", environment, |ctx| {
+        ctx.when("not having added any items", |ctx| {
+            ctx.then("it is empty", |env| assert!(env.set.is_empty()));
         });
 
-        ctx.describe("rspec can use bools", |ctx| {
+        ctx.when("adding an new item", |ctx| {
+            ctx.before_all(|env| {
+                env.len_before = env.set.len();
+                env.set.insert(42);
+            });
 
-            ctx.it("should pass if true", || true);
+            ctx.then("it is not empty any more", |env| {
+                assert!(!env.set.is_empty());
+            });
 
-            ctx.it("should fail if false", || false);
+            ctx.then("its len increases by 1", move |env| {
+                assert_eq!(env.set.len(), env.len_before + 1);
+            });
 
-            ctx.it("is convenient for comparisons", || {
-                (42 % 37 + 2) > 3
-            })
-        });
+            ctx.when("adding it again", |ctx| {
+                ctx.before_all(|env| {
+                    env.len_before = env.set.len();
+                    env.set.insert(42);
+                });
 
-        ctx.describe("rspec can use units", |ctx| {
+                ctx.then("its len remains the same", move |env| {
+                    assert_eq!(env.set.len(), env.len_before);
+                });
+            });
+        });
 
-            ctx.it("should pass if the return is ()", || {});
+        ctx.when("returning to outer context", |ctx| {
+            ctx.then("it is still empty", |env| assert!(env.set.is_empty()));
+        });
 
-            ctx.it("is convenient for asserts", || assert_eq!(1, 1));
+        ctx.then("panic!(…) fails", move |_env| {
+            panic!("Some reason for failure.")
```

Comment:
explain the panic usage. It may not be relevant in this file, but in another example